### PR TITLE
device: allow NULL init function 

### PR DESCRIPTION
**Pull request by gmarull**
Some devices do not need to perform any initialization, so allow the
init function to be NULL. In this case, the initialization code will
just mark the device as initialized, i.e. ready.

Removed all redundant init functions I found throughout the tree (2nd commit).

### CODE REVIEW

**Review by de-nordic — DISMISSED** [review]

File: kernel/init.c [hunk 1]
```diff
@@ -252,20 +252,26 @@ static void z_sys_init_run_level(enum init_level level)
...
+					}
+					dev->state->init_res = rc;
 				}
-				if (rc > UINT8_MAX) {
-					rc = UINT8_MAX;
-				}
-				dev->state->init_res = rc;
 			}
+
 			dev->state->initialized = true;
```
Comment (henrikbrixandersen) [inline kernel/init.c:273]: Would it be possible to ignore the initialisation level completely for devices without an initialisation function - and just statically mark the as "initialised"?

**Review by stephanosio — DISMISSED** [review]

File: include/zephyr/drivers/i2c.h [hunk 2]
```diff
@@ -539,7 +539,11 @@ static inline void i2c_xfer_stats(const struct device *dev, struct i2c_msg *msgs
 		stats_init(&state->stats.s_hdr, STATS_SIZE_32, 4,	\
 			   STATS_NAME_INIT_PARMS(i2c));			\
 		stats_register(dev->name, &(state->stats.s_hdr));	\
-		return init_fn(dev);					\
+		if (init_fn != NULL) {					\
+			return init_fn(dev);				\
+		}							\
+									\
+		return 0;						\
```
Comment (henrikbrixandersen) [inline include/zephyr/drivers/i2c.h:546]: A similar change is needed in `include/zephyr/drivers/can.h`.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

File: kernel/init.c (hunk 1 above)
Comment (gmarull) [reply kernel/init.c:273]: Hmm, how would you do that?

**Comment by gmarull** [comment]
@galak I don't think this really applies to treewide, since this is a non-breaking change. I just took the time to clean up redundant init calls, but they could really stay without any impact.

**Comment by gmarull** [comment]
> @galak I don't think this really applies to treewide, since this is a non-breaking change. I just took the time to clean up redundant init calls, but they could really stay without any impact.

Ref.
```
A *treewide change* is defined as any change to Zephyr APIs, coding practices,
or other development requirements that either implies required changes
throughout the zephyr source code repository or can reasonably be expected to
do so for a wide class of external Zephyr-based source code.
```

**Review by de-nordic — APPROVED** [review]

**Review by stephanosio — APPROVED** [review]

**Review by henrikbrixandersen — APPROVED** [review]

File: kernel/init.c (hunk 1 above)
Comment (evgeniy-paltsev) [reply kernel/init.c:273]: > @gmarull Hmm, how would you do that?

Probably in Z_DEVICE_DEFINE we can set state in Z_DEVICE_BASE_DEFINE and we can drop creating init entry with Z_DEVICE_INIT_ENTRY_DEFINE.

However I'm not sure if we can integrate it in existing Z_DEVICE_DEFINE macro without defining new one for the devices which doesn't require initialization.

Comment (gmarull) [reply kernel/init.c:273]: The problem I see is that we cannot evaluate at compile time if `init_fn` is `NULL` and make a decision based on that. We should probably introduce a new macro like `DEVICE_DEFINE_NOINIT`, that skips the creation of a device init entry and sets initialized = true.

Comment (mbolivar-nordic) [reply kernel/init.c:273]: It's probably not worth it to do this at link time after the first stage link, is it?

**Pull request merged**